include/wil/stl.h

```diff
@@ -172,6 +172,18 @@ namespace wil
         }
 
     private:
+        // Bounds-checked version of char_traits::length, like strnlen.
+        static constexpr size_type length_n(_In_reads_opt_(length) const TChar* str, size_type buf_size) noexcept

```

Comment:
I'm always hesitant of custom writing things like this. Using something like `strnlen` directly would obviously be the most preferable, but then you'd need to overload on `TChar`, etc... `if constexpr` wouldn't be _that_ bad here. Otherwise, we can construct `basic_string_view` using `stringArrayLength` and then resize/fail-fast based off the result of `find_first_of`.

Reply:
`TChar` overloads are likely not to work on u16/u32strings.

Reply:
I was more suggesting specializing on types that do have a function/intrinsic available. E.g. `char` and `wchar_t` and then fall back to something like this for other types.

FWIW the `wchar_t` <-> `char16_t` cast is used pretty freely in things like C++/WinRT

Reply:
I'll check it, with one remark - I doubt I can use `if constexpr` since that'll lock the code to C++17 and newer. I'll have to look into SFINAE.

Reply:
`basic_string_view` was also added in C++17 (I suppose your edit suggests you realized that). But I also forgot that the string functions didn't get the `constexpr` update when `char_tratits` etc. did :/. I'm also unsure if `WI_STL_FAIL_FAST_IF` is `constexpr`, but that's to look at another time.

`find_first_of` seems reasonable IMO and is no more complex than what you have here.
```c++
        // Bounds-checked version of char_traits::length, like strnlen.
        static constexpr size_type length_n(_In_reads_opt_(length) const TChar* str, size_type length) noexcept
        {
            std::basic_string_view<TChar> view(str, length);
            auto pos = view.find_first_of(TChar());
            return (pos == view.npos) ? length : pos; // Perhaps maybe even move the failfast here to be simpler
        }
```
I'd probably personally prefer doing all the operations on `this`, but alas there is no way to change the length of a `basic_string_view` apart from assignment (e.g. with `.substr()`), but that likely causes a mess of other issues...

Reply:
Ah, like this, good idea! Just one note, though - fast failing inside this function discards a possibility that there is a null terminator right after the passed size, as per my footnote in the initial post. Should this be a supported scenario (that then disqualifies true `constexpr`), or should it be forbidden? If the latter, then the test checking for this particular scenario should be discarded.

Reply:
Given that this is just used for construction with arrays, I'm not sure that concerning ourselves with the data outside of the bounds of the array is a good idea, and I'd like to believe that nobody has taken a dependency on it. E.g. maybe it's a value that's initialized to zero, but later changes to something else. There's also no guarantee that the byte following the array can even be read. E.g. it's possible to get a page fault/access violation. Are you saying that there's an existing test that expects to be able to read one past the end of the array and find a null character? I only see the tests you added which verify a crash (which I'd expect).

Of course, you also have the issue that there's no guarantee of `const`-ness of the array, so someone could pass a non-`const` array that they later change breaking the precondition, but I'm not really worried about people going to such lengths to shoot themselves in the foot.

Reply:
> Are you saying that there's an existing test that expects to be able to read one past the end of the array and find a null character?

Technically no, the passing test I added does, though, hence the remark `Overloads taking an explicit length trust the user that they ensure valid memory follows the buffer`.

If you think it's worth taking the "risk" here and disallowing it entirely, I'm all for it. Disallowing the assumption that a null terminator follows the string would probably simplify code, allow those string views to be truly `constexpr` (as reading past array bounds cannot be), and probably be safer overall. My current passing test would become invalid but I could add another one, e.g. with a `{`a`, `b`, `c`, `\0`, `b`, `u`, `f`,}` array where the expectation is that it produces an `"abc"` view.

Reply:
Ahh, okay, I misread. I thought you were referring to constructing with arrays, my bad. Yes, I agree, the caller should assume that it's safe to read the null terminator in all situations. E.g. if I pass `(ptr, length)`, then it should be assumed that reading from `ptr[length]` is valid, won't AV, and gives the null character.

Reply:
Indeed, I was just editing my answer as you left yours, here it is unedited:

```
EDIT:
On a second look, my "passing" tests verify another constructor, but that is potentially alright. Securing this specific constructor against reading past the array bounds would change the contract, so it might be better to trust the user that they pass a valid size.
```

Looks like we effectively agreed on the same! My "passing" test assumes this exactly, i.e. that `ptr[length] == '\0`.

Reply:
Yeah, I just mis-understood your question. The tests as you have them written make sense to me and look correct. If I try and construct with an array, I expect the null character to be within the bounds of the array (otherwise failfast). If I construct with a `(pointer, length)` pair, I'm guaranteeing that it's safe to read from index `0` to index `length` and that `pointer[length]` is the null character. If constructing with a `(pointer, length)` pair does not produce a constant expression (e.g. would read past the end of an array), that's on the caller and they therefore should not try and do something like that in a `constexpr` context.

Reply:
There, I applied the changes we agreed upon. `WI_STL_FAIL_FAST_IF` is **not** `constexpr`, but the compiler realizes that the code branch containing it is never taken in the current `constexpr` contexts, and thus it's fine. Had it been `WI_STL_FAIL_FAST_IF(pos == view.npos);`, it would not compile.